hedera-node/hedera-app/src/main/java/com/hedera/node/app/authorization/PrivilegesVerifier.java

```diff
@@ -0,0 +1,217 @@
+
+    private final AccountsConfig accountsConfig;
+    private final FilesConfig filesConfig;
+    private final long numReservedSystemEntities;
+
+    @Inject
+    public PrivilegesVerifier(@NonNull final ConfigProvider configProvider) {
+        requireNonNull(configProvider, "configProvider cannot be null");
+
+        final var configuration = configProvider.getConfiguration();
+        this.accountsConfig = configuration.getConfigData(AccountsConfig.class);
+        this.filesConfig = configuration.getConfigData(FilesConfig.class);
+        this.numReservedSystemEntities =
+                configuration.getConfigData(LedgerConfig.class).numReservedSystemEntities();
+    }
+
+    /**
+     * Checks whether an account is exempt from paying fees.
+     *
+     * @param accountID the {@link AccountID} to check
+     * @param functionality the {@link HederaFunctionality} of the transaction
+     * @param txBody the {@link TransactionBody} of the transaction
+     * @return {@code true} if the account is exempt from paying fees, otherwise {@code false}
+     */
+    public SystemPrivilege hasPrivileges(
+            @NonNull final AccountID accountID,
+            @NonNull final HederaFunctionality functionality,
+            @NonNull final TransactionBody txBody) {
+        return switch (functionality) {
+                // Authorization privileges for special transactions
+            case FREEZE -> checkFreeze(accountID);
+            case SYSTEM_DELETE -> checkSystemDelete(accountID, txBody.systemDeleteOrThrow());
+            case SYSTEM_UNDELETE -> checkSystemUndelete(accountID, txBody.systemUndeleteOrThrow());
+            case UNCHECKED_SUBMIT -> checkUncheckedSubmit(accountID);
+
+                // Authorization privileges for file updates and appends
+            case FILE_UPDATE -> checkFileChange(
+                    accountID, txBody.fileUpdateOrThrow().fileIDOrThrow().fileNum());
+            case FILE_APPEND -> checkFileChange(
+                    accountID, txBody.fileAppendOrThrow().fileIDOrThrow().fileNum());
+            case CONTRACT_UPDATE -> checkFileChange(
+                    accountID,
+                    txBody.contractUpdateInstanceOrThrow().contractIDOrThrow().contractNumOrThrow());
+
+                // Authorization for crypto updates
+            case CRYPTO_UPDATE -> checkCryptoUpdate(accountID, txBody.cryptoUpdateAccountOrThrow());
+
+                // Authorization for deletes
+            case FILE_DELETE -> checkEntityDelete(
+                    txBody.fileDeleteOrThrow().fileIDOrThrow().fileNum());
+            case CRYPTO_DELETE -> checkEntityDelete(
+                    txBody.cryptoDeleteOrThrow().deleteAccountIDOrThrow().accountNumOrThrow());
+            case CONTRACT_DELETE -> checkEntityDelete(
+                    txBody.contractDeleteInstanceOrThrow().contractIDOrThrow().contractNumOrThrow());
+
+            default -> SystemPrivilege.UNNECESSARY;
+        };
+    }
+
```

Comment:
Will be good to have javadocs on all these methods, not a blocker for the PR.

Reply:
Good idea, will add it.